src/dspbase.jl

```diff
@@ -117,24 +117,58 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
     filt(b, a, x)
 end
 
+# padded must start at index 1, but u can have arbitrary offset
 function _zeropad!(padded::AbstractVector, u::AbstractVector)
-    ulen = length(u)
-    padlen = length(padded)
-    copyto!(padded, 1, u, 1, ulen)
-    padded[ulen + 1:padlen] .= 0
+    datasize = length(u)
+    # Use axes to accommodate arrays that do not start at index 1
+    padsize = length(padded)
+    data_first_i = first(axes(u, 1))
+    copyto!(padded, 1, u, data_first_i, datasize)
+    padded[1 + datasize:padsize] .= 0
     padded
 end
-function _zeropad!(padded::AbstractArray{<:Any, N}, u::AbstractArray{<:Any, N}) where N
-    datainds = CartesianIndices(u)
-    copyto!(padded, datainds, u, datainds)
+
+# padded must start at index 1, but u can have arbitrary offset
+function _zeropad!(padded::AbstractArray{<:Any, N},
+                   u::AbstractArray{<:Any, N}) where N
+    # Copy the data to the beginning of the padded array
+    padsize = size(padded)
+    pad_ax = axes(padded)
+    datasize = size(u)
+    pad_data_ranges = range.(1, datasize)
+    copyto!(padded, CartesianIndices(pad_data_ranges), u, CartesianIndices(u))
+
+    # Step through each dimension, and fill the trailing indices in that
+    # dimension with zeros.
+    #
+    # This is accomplished by going from the last to the first dimension, and
+    # for each dimension, finding the biggest rectangular region that needs to
+    # be zero padded. This region corresponds to the  trailing indices after the
+    # data in the selected dimension, all the indices in lower dimensions, and
+    # the same indices as the data for higher dimensions, corresponding to
+    # regions that were missed in that dimension. This should allow column-major
+    # friendly access to the array.
+    #
+    # Geometric intuition for two dimensions:
+    #   1 1 1 0 0     1 1 1    0 0
+    #   1 1 1 0 0     1 1 1    0 0
+    #   1 1 1 0 0  =  1 1 1    0 0
+    #   0 0 0 0 0              0 0
+    #   0 0 0 0 0     0 0 0    0 0
+    #                 0 0 0
+
     pad_ranges = Vector{UnitRange{Int}}(undef, N)
-    pad_ranges .= axes(padded)
+    pad_ranges .= pad_ax # Initially select the entire dimension
     for i = N:-1:1
-        nu = size(u, i)
-        pad_ranges[i] = nu + 1:size(padded, i)
-        padinds = CartesianIndices(NTuple{N, UnitRange{Int}}(pad_ranges))
-        padded[padinds] .= 0
-        pad_ranges[i] = 1:nu
+        # Trailing indices for this dimension
+        pad_ranges[i] = datasize[i] + 1 : padsize[i]
+
+        # Make the rectangular region and set it to zero
+        pad_region = CartesianIndices(NTuple{N, UnitRange{Int}}(pad_ranges))
+        padded[pad_region] .= 0
+
+        # Set this dimension to be just the portion that was missed (same as data)
+        pad_ranges[i] = pad_data_ranges[i]
```

Comment:
Zeroing out large contiguous blocks of memory should be pretty fast, and when `u` and `v` are similarly sized and N-dimensional only `1/2^N` of the entries are going to be nonzero. I did a little benchmarking against a version that just prefills `padded` with zeros:

```julia
using BenchmarkTools
using DSP

function zeropad_prefill!(padded::AbstractArray{<:Any, N},
                          u::AbstractArray{<:Any, N}) where N
    # Copy the data to the beginning of the padded array
    fill!(padded, zero(eltype(padded)))
    datasize = size(u)
    pad_data_ranges = range.(1, datasize)
    copyto!(padded, CartesianIndices(pad_data_ranges), u, CartesianIndices(u))
    
    padded
end

padded = rand(1000, 1000)
u = rand(500, 500)

display(@benchmark DSP._zeropad!(padded, u))
display(@benchmark zeropad_prefill!(padded, u))
```

On my system the `prefill` version is faster until `u` gets to be about `900x900`